cpp/src/arrow/dataset/scanner.cc

```diff
@@ -58,6 +55,62 @@ static ScanTaskIterator GetScanTaskIterator(DataFragmentIterator fragments,
   return MakeFlattenIterator(std::move(maybe_scantask_it));
 }
 
+template <typename Fn>
```

Comment:
I think `MapScanTask` is an unclear level of lambdas and mappings. Please refactor to a `FilteredAndProjectedScanTask` which holds a pointer to the evaluator, expression, and projector and constructs the filtered and projected record batch iterator in `Scan()`.